### PR TITLE
Add Conditional Inputs for _GeneratePackageManagerJava and _GenerateJavaStubs

### PR DESCRIPTION
**Pull request by dellis1972**
This commit updates the inputs for `_GenerateJavaStubs` and `_GeneratePackageManagerJava` so that the `@(AndroidEnvironment)` and `@(LibraryEnvironments)` items are only included if we are building a release build. Or a build which embeds the assemblies into the apk. 

This is so that we can evantually fast deploy these files to the device via https://github.com/xamarin/monodroid/pull/1547. 
A new unit test has been added to make sure that the environment values can be used correctly on device.

### CODE REVIEW

**Comment by dellis1972** [comment]
needs https://github.com/xamarin/monodroid/pull/1547

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 1]
```diff
@@ -1684,12 +1696,23 @@ because xbuild doesn't support framework reference assemblies.
...
 
+<Target Name="_GetGeneratePackageManagerJavaInputs">
+  <ItemGroup>
+    <_GeneratePackageManagerJavaInputs Include="@(_AndroidMSBuildAllProjects)" />
+    <_GeneratePackageManagerJavaInputs Include="$(_ResolvedUserAssembliesHashFile)" />
+    <_GeneratePackageManagerJavaInputs Include="@(_ResolvedUserMonoAndroidAssemblies)" />
+    <_GeneratePackageManagerJavaInputs Include="$(_AndroidBuildPropertiesCache)" />
+    <_GeneratePackageManagerJavaInputs Include="@(AndroidEnvironment);@(LibraryEnvironments)" Condition=" '$(EmbedAssembliesIntoApk)' == 'true' " />
+  </ItemGroup>
+</Target>
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:1711]: Can both targets use the same item group?

I'm thinking about:
* https://github.com/dotnet/android/pull/9174

Where there was a bug if `_GenerateJavaStubs` runs and `_GeneratePackageManagerJava` doesn't.

Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:1711]: Or maybe that was vice versa.

Comment (dellis1972) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:1711]: There is still the Manifest which is different between the two. But maybe one could "use" the other and add the manifest.

**Review by jonathanpeppers — APPROVED** [review]

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 2]
```diff
@@ -1446,12 +1446,24 @@ because xbuild doesn't support framework reference assemblies.
...
 </PropertyGroup>
 
+<Target Name="_GetGenerateJavaStubsInputs">
+  <ItemGroup>
+    <_GenerateJavaStubsInputs Include="@(_AndroidMSBuildAllProjects)" />
+    <_GenerateJavaStubsInputs Include="$(_ResolvedUserAssembliesHashFile)" />
+    <_GenerateJavaStubsInputs Include="@(_ResolvedUserMonoAndroidAssemblies)" />
+    <_GenerateJavaStubsInputs Include="$(_AndroidManifestAbs)" />
+    <_GenerateJavaStubsInputs Include="$(_AndroidBuildPropertiesCache)" />
+    <_GenerateJavaStubsInputs Include="@(AndroidEnvironment);@(LibraryEnvironments)" Condition=" '$(EmbedAssembliesIntoApk)' == 'true' " />
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:1460]: If we introduce a private flag `$(_AndroidFastDeployEnvironmentFiles)` for https://github.com/xamarin/monodroid/pull/1547, could we also check it here?

`@(_GenerateJavaStubsInputs)` should contain env var files if `$(_AndroidFastDeployEnvironmentFiles)` is `false`.

Comment (dellis1972) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:1460]: How is this different for `EmbedAssembliesIntoApk` unless the idea is to allow us to override the behaviour?

Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:1460]: `$(_AndroidFastDeployEnvironmentFiles)` would let you keep using Fast Deployment, and just not Fast Deploy the environment files. It's just an extra a safety flag if something goes wrong.

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 3]
```diff
@@ -288,6 +288,10 @@ Copyright (C) 2011-2012 Xamarin. All rights reserved.
 	<_NativeAssemblySourceDir>$(IntermediateOutputPath)android\</_NativeAssemblySourceDir>
 	<_AndroidUseNewTypemaps>True</_AndroidUseNewTypemaps>
 	<_SkipJniAddNativeMethodRegistrationAttributeScan Condition=" '$(_SkipJniAddNativeMethodRegistrationAttributeScan)' == '' ">False</_SkipJniAddNativeMethodRegistrationAttributeScan>
+
+	<_AndroidFastDeployEnvironmentFiles Condition=" '$(_AndroidFastDeployEnvironmentFiles)' == '' And '$(EmbedAssembliesIntoApk)' == 'true' ">True</_AndroidFastDeployEnvironmentFiles>
+	<_AndroidFastDeployEnvironmentFiles Condition=" '$(_AndroidFastDeployEnvironmentFiles)' == '' ">False</_AndroidFastDeployEnvironmentFiles>
+
```
Comment (dellis1972) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:294]: @jonathanpeppers Was this what you had in mind? It defaults to the value of `EmbedAssembliesIntoApk` but will let us override it.

**Comment by jonpryor** [comment]
Draft Commit Message:

```markdown
[Xamarin.Android.Build.Tasks] Add Inputs for _GenerateJavaStubs (#9319)

Context: https://github.com/xamarin/monodroid/pull/1547

This commit updates the Inputs for the `_GenerateJavaStubs` and
`_GeneratePackageManagerJava` targets so that the
`@(AndroidEnvironment)` and `@(LibraryEnvironments)` items are only
included when we are building a release build, or a build which
embeds the assemblies into the apk.

This is so that we can eventually fast deploy these files to the
device; see also xamarin/monodroid#1547. 

A new unit test has been added to make sure that the environment
values can be used correctly on device.

```

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 4]
```diff
@@ -288,6 +288,10 @@ Copyright (C) 2011-2012 Xamarin. All rights reserved.
 	<_NativeAssemblySourceDir>$(IntermediateOutputPath)android\</_NativeAssemblySourceDir>
 	<_AndroidUseNewTypemaps>True</_AndroidUseNewTypemaps>
 	<_SkipJniAddNativeMethodRegistrationAttributeScan Condition=" '$(_SkipJniAddNativeMethodRegistrationAttributeScan)' == '' ">False</_SkipJniAddNativeMethodRegistrationAttributeScan>
+
+	<_AndroidFastDeployEnvironmentFiles Condition=" '$(_AndroidFastDeployEnvironmentFiles)' == '' And '$(EmbedAssembliesIntoApk)' == 'true' ">True</_AndroidFastDeployEnvironmentFiles>
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:292]: @dellis1972: This `Condition` feels backwards to me? `$(_AndroidFastDeployEnvironmentFiles)` reads as "affirmative", as in "if `$(_AndroidFastDeployEnvironmentFiles)` is true, then `@(AndroidEnvironment)`/etc. should *also* be fast deployed".

Yet, `$(_AndroidFastDeployEnvironmentFiles)` will only default to True if `$(EmbedAssembliesIntoApk)` is true, which means fast deployment is *disabled*.

I don't think I understand the semantics of this property. :-(

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 5]
```diff
@@ -1446,12 +1450,24 @@ because xbuild doesn't support framework reference assemblies.
...
 </PropertyGroup>
 
+<Target Name="_GetGenerateJavaStubsInputs">
+  <ItemGroup>
+    <_GenerateJavaStubsInputs Include="@(_AndroidMSBuildAllProjects)" />
+    <_GenerateJavaStubsInputs Include="$(_ResolvedUserAssembliesHashFile)" />
+    <_GenerateJavaStubsInputs Include="@(_ResolvedUserMonoAndroidAssemblies)" />
+    <_GenerateJavaStubsInputs Include="$(_AndroidManifestAbs)" />
+    <_GenerateJavaStubsInputs Include="$(_AndroidBuildPropertiesCache)" />
+    <_GenerateJavaStubsInputs Include="@(AndroidEnvironment);@(LibraryEnvironments)" Condition=" '$(_AndroidFastDeployEnvironmentFiles)' == 'false' " />
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:1464]: I believe the advice was that one shouldn't compare `== 'false'`, but instead should use `!= 'true'`.

**Comment by dellis1972** [comment]
I found a bug, don't merge it.

**Review by grendello — APPROVED** [review]

**Comment by jonpryor** [comment]
@dellis1972: [`FastDeployEnvironmentFiles(False)` is failing](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=10287464&view=ms.vss-test-web.build-test-results-tab&runId=111624334&resultId=100000&paneView=debug):

> The Environment variable "Foo" was not set.
> Expected: String containing "Foo=Bar"
> But was:  "--------- beginning of main …

I've restarted the tests just in case it's sporadic…

**Comment by dellis1972** [comment]
> @dellis1972: [`FastDeployEnvironmentFiles(False)` is failing](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=10287464&view=ms.vss-test-web.build-test-results-tab&runId=111624334&resultId=100000&paneView=debug):
> 
> > The Environment variable "Foo" was not set.
> > Expected: String containing "Foo=Bar"
> > But was:  "--------- beginning of main …
> 
> I've restarted the tests just in case it's sporadic…

This test is going to fail untl the monodroid side is merged.

**Pull request merged**